Deferred/Deferred.swift

```diff
@@ -6,39 +6,131 @@
+        withUnsafeMutablePointerToElements { elementsPtr in
+            elementsPtr.initialize(box)
+        }
+    }
+    
+    func withValue(body: Value -> Void) {
+        withUnsafeMutablePointerToElements { elementsPtr in
+            guard let box = elementsPtr.memory else { return }
+            body(box.value)
+        }
+    }
+    
+    func fill(value: Value, onFill: SideData -> Void) -> Bool {
+        let box = Storage(value)
+        return withUnsafeMutablePointers { (valuePtr, elementsPtr) in
+            guard compareAndSwap(old: nil, new: box, to: elementsPtr) else { return false }
+            onFill(valuePtr.memory)
+            return true
+        }
+    }
+    
+}
+
+// A generic catch-all queue for when you just want to throw some work into the
+// concurrent pile. As an alternative to the `QOS_CLASS_UTILITY` global queue,
+// work dispatched onto this queue matches the QoS of the caller, which is
+// generally the right behavior.
+//
+// The technique is described and used in Core Foundation:
+// http://opensource.apple.com/source/CF/CF-1153.18/CFInternal.h
+private var genericQueue: dispatch_queue_t! {
+    return dispatch_get_global_queue(qos_class_self(), 0)
 }
 
 /// A deferred is a value that may become determined (or "filled") at some point
 /// in the future. Once a deferred value is determined, it cannot change.
 public struct Deferred<Value> {
-    typealias UponBlock = (dispatch_queue_t, Value -> ())
-    private typealias Protected = (protected: Value?, uponBlocks: [UponBlock])
     
-    private var protected: LockProtected<Protected>
+    private var storage = DeferredBuffer<Value, dispatch_block_t>.create {
+        dispatch_block_create(DISPATCH_BLOCK_NO_QOS_CLASS, {})
+    }
+
+    private func upon(queue: dispatch_queue_t, options inFlags: dispatch_block_flags_t, body: Value -> Void) -> dispatch_block_t {
+        let flags = dispatch_block_flags_t(inFlags.rawValue | DISPATCH_BLOCK_ASSIGN_CURRENT.rawValue)
+        let block = dispatch_block_create(flags) {
+            self.storage.withValue(body)
+        }
+        dispatch_block_notify(storage.value, queue, block)
+        return block
+    }
     
-    private init(_ value: Value?) {
-        protected = LockProtected(item: (value, []))
+    private func markFilled(onFilled: dispatch_block_t) {
+        dispatch_block_cancel(onFilled)
+        onFilled()
     }
+
+    // MARK: -
     
     /// Initialize an unfilled Deferred.
     public init() {
-        self.init(nil)
+        storage.initializeWith(nil)
     }
     
     /// Initialize a filled Deferred with the given value.
     public init(value: Value) {
-        self.init(value)
+        storage.initializeWith(value)
+        markFilled(storage.value)
     }
-    
+
     /// Check whether or not the receiver is filled.
     public var isFilled: Bool {
-        return protected.withReadLock { $0.protected != nil }
+        return dispatch_block_testcancel(storage.value) != 0
```

Comment:
This is another example of the same thing as my last comment about more documentation about the general approach.

I look at the line that was replaced, and I can make sense of it.

Here is my naive thought process as I take a look at this method for the first time, with the old line:

Before I even open the code I think:

1. I bet this method will just look at whether or not is has a value stored yet for the private property that it must manage, that property must be an optional.
2. I haven't thought about all of the scenarios, but I bet there are some threading concerns there, so I wouldn't be surprised to see something related to threading.
3. I open the code and look at the method.
4. `protected.withReadLock` ... ok, that makes some sense ... must be dealing with some threading concerns.
5. I don't know exactly what this `protected` thing is, but it appears to be dealing with locking/threads so that's enough insight for me.
5. Yep .. then it checks the optional value to see if it is nil or not ... ok, makes sense to me.

(The above is just what goes on in my mind as a potential maintainer of this code.)

When I look at the new code, I still have the same thoughts #1-3 as I am about to read the code, but then I look at the code and just feel like "what did I miss?".  "Maybe I am looking at the wrong thing?"

Again, I am not at all suggesting that the code is incorrect.  I am simply concerned that it is more difficult to understand the motivation.  The earlier suggestion of more documentation might help ease the maintenance of this class.